### PR TITLE
[doc] Add more reference around predict_step

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds more context on when and how to use ``predict_step`` along ``trainer.predict``

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -1115,6 +1115,25 @@ def predict_step(self, batch: Any, batch_idx: int, dataloader_idx: Optional[int]
         By default, it calls :meth:`~pytorch_lightning.core.lightning.LightningModule.forward`.
         Override to add any processing logic.
 
+        The :meth:`~pytorch_lightning.core.lightning.LightningModule.predict_step` is used
+        to scale inference on multi-devices.
+
+        To prevent an OOM error, it is possible to use :class:`~pytorch_lightning.callbacks.BasePredictionWriter`
```
Comment (kaushikb11) [inline pytorch_lightning/core/lightning.py:1121]: Could also add, predictions are not returned for Spawn Training Plugins.

Comment (justusschock) [reply pytorch_lightning/core/lightning.py:1121]: @kaushikb11 I wonder if this is something we might want to change with the approach from #7916 .

File: docs/source/common/lightning_module.rst [hunk 2]
```diff
@@ -554,12 +563,20 @@ Inference in research
 ^^^^^^^^^^^^^^^^^^^^^
 In the case where we want to perform inference with the system we can add a `forward` method to the LightningModule.
 
+.. note:: When using forward, you are responsible to call :func:`~torch.nn.Module.eval` and use the :func:`~torch.no_grad` decorator
```
Comment (awaelchli) [inline docs/source/common/lightning_module.rst:566]:
```suggestion
.. note:: When using forward, you are responsible to call :func:`~torch.nn.Module.eval` and use the :func:`~torch.no_grad` context manager
```

**Review by awaelchli — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/core/lightning.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/core/lightning.py:1121]: We could. It was why we introduced this :)

Comment (tchaton) [reply pytorch_lightning/core/lightning.py:1121]: Done !

File: docs/source/common/lightning_module.rst (hunk 2 above)
Comment (tchaton) [reply docs/source/common/lightning_module.rst:566]: Done !

File: pytorch_lightning/core/lightning.py [hunk 3]
```diff
@@ -1115,6 +1115,29 @@ def predict_step(self, batch: Any, batch_idx: int, dataloader_idx: Optional[int]
...
+        callback to write the predictions to disk or database after each batch or on epoch end.
+
+        The :class:`~pytorch_lightning.callbacks.BasePredictionWriter` should be used when using a spawn
+        based accelerator. This happens whe running ``Trainer(accelerator="ddp_spawn")``
```
Comment (kaushikb11) [inline pytorch_lightning/core/lightning.py:1125]:
```suggestion
        The :class:`~pytorch_lightning.callbacks.BasePredictionWriter` should be used while using a spawn
        based accelerator. This happens for ``Trainer(accelerator="ddp_spawn")``
```

**Review by kaushikb11 — APPROVED** [review]
